### ISSUE TITLE
Middleware unable to write headers in before blocks

### ISSUE THREAD

**Issue by yukimono** [ruby-grape/grape#1069]
Reproduction:
```ruby
class API < Grape::API
  before do
    header "X-Test-A", "Success"
  end
  
  class HeaderTest < Grape::Middleware::Base
    def before
      endpoint = env['api.endpoint']
      endpoint.header "X-Test-B", "Success"
    end
  end

  use HeaderTest

  get '/' do
    present ""
  end
end
```

Expected result:

```
  HTTP/1.1 200 OK
  Content-Type: text/plain
  Content-Length: 0
  Cache-Control: max-age=0, private, must-revalidate
  Connection: keep-alive
  X-Test-A: Success
  X-Test-B: Success
  Server: thin
```
Actual result:

```
  HTTP/1.1 500 Internal Server Error
  Content-Type: text/plain
  Content-Length: 39
  Cache-Control: max-age=0, private, must-revalidate
  Connection: keep-alive
  Server: thin

  undefined method `[]=' for nil:NilClass
```

Cause: ~~Headers are sent before any `before` calls in middleware~~ Before is called before any response is created.

Workaround: Set headers in `after` (`@app_response[1]`)

**Comment by yukimono** [comment]
N.B. trying to access `response` also yields ``undefined method `[]' for nil:NilClass``

**Comment by dblock** [comment]
I agree it should be possible, marking this as a feature request.

**Comment by gottfrois** [comment]
I also encounter the exact same issue. I was expecting to be able to add custom headers in the before method but the headers instance variable is defined after. Looking forward for this to be solved :)

**Comment by namusyaka** [comment]
I'd like to fix this after solving #1243

**Comment by namusyaka** [comment]
@dblock This can be closed by #1244.
